cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller_test.go

```diff
@@ -632,168 +619,168 @@ func TestControllerNodeGroupForNodeWithPositiveScalingBounds(t *testing.T) {
+		t.Fatalf("expected an error")
+	}
+
+	// Test #8: machinedeployment with bad scaling bounds results in an error and no nodegroups
+	machineDeploymentConfigs = mustCreateMachineDeploymentTestConfigs("MachineDeployment", 2, 1, 1, annotations)
+	if err := addTestConfigs(t, controller, machineDeploymentConfigs...); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if _, err := controller.nodeGroups(); err == nil {
+		t.Fatalf("expected an error")
+	}
 }
 
-func TestControllerNodeGroupForNodeLookup(t *testing.T) {
+func TestControllerNodeGroupsNodeCount(t *testing.T) {
 	type testCase struct {
-		description    string
-		annotations    map[string]string
-		lookupSucceeds bool
+		nodeGroups    int
+		nodesPerGroup int
 	}
 
 	var testCases = []testCase{{
-		description:    "lookup is nil because no annotations",
-		annotations:    map[string]string{},
-		lookupSucceeds: false,
+		nodeGroups:    0,
+		nodesPerGroup: 0,
 	}, {
-		description: "lookup is nil because scaling range == 0",
-		annotations: map[string]string{
-			nodeGroupMinSizeAnnotationKey: "1",
-			nodeGroupMaxSizeAnnotationKey: "1",
-		},
-		lookupSucceeds: false,
+		nodeGroups:    1,
+		nodesPerGroup: 0,
 	}, {
-		description: "lookup is successful because scaling range >= 1",
-		annotations: map[string]string{
-			nodeGroupMinSizeAnnotationKey: "1",
-			nodeGroupMaxSizeAnnotationKey: "2",
-		},
-		lookupSucceeds: true,
+		nodeGroups:    2,
+		nodesPerGroup: 10,
 	}}
 
-	test := func(t *testing.T, tc testCase, testObjs *clusterTestConfig, node *apiv1.Node) {
-		controller, stop := testObjs.newMachineController(t)
+	test := func(t *testing.T, tc testCase, testConfigs []*testConfig) {
+		controller, stop := mustCreateTestController(t, testConfigs...)
 		defer stop()
 
-		ng, err := controller.nodeGroupForNode(node)
+		nodegroups, err := controller.nodeGroups()
 		if err != nil {
 			t.Fatalf("unexpected error: %v", err)
 		}
-
-		if ng == nil && tc.lookupSucceeds {
-			t.Fatalf("expected non-nil from lookup")
+		if l := len(nodegroups); l != tc.nodeGroups {
+			t.Fatalf("expected %d, got %d", tc.nodeGroups, l)
 		}
 
-		if ng != nil && !tc.lookupSucceeds {
-			t.Fatalf("expected nil from lookup")
-		}
-
-		if !tc.lookupSucceeds {
-			return
-		}
-
-		var expected string
-
-		if testObjs.machineDeployment != nil {
-			expected = path.Join(testObjs.machineDeployment.Namespace, testObjs.machineDeployment.Name)
-		} else {
-			expected = path.Join(testObjs.machineSet.Namespace, testObjs.machineSet.Name)
+		for i := range nodegroups {
+			nodes, err := nodegroups[i].Nodes()
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if l := len(nodes); l != tc.nodesPerGroup {
+				t.Fatalf("expected %d, got %d", tc.nodesPerGroup, l)
+			}
 		}
+	}
 
-		if actual := ng.Id(); actual != expected {
-			t.Errorf("expected %q, got %q", expected, actual)
-		}
+	annotations := map[string]string{
+		nodeGroupMinSizeAnnotationKey: "1",
+		nodeGroupMaxSizeAnnotationKey: "2",
```

Comment:
You are removing `(1,1)` and `(1,2)` min/max annotations and replacing them with `(1,2)` for all test cases. Is that OK?

Reply:
Yes - this test has changed and now requires the bounds to be > 1 so that `nodeGroups()` will actually return a value. If the bounds were left at 1 we have this check in `machineSetNodeGroups()` and `machineDeploymentNodeGroups()`.

```go
// add nodegroup iff it has the capacity to scale
if ng.MaxSize()-ng.MinSize() > 0 {
	nodegroups = append(nodegroups, ng)
}
```